File: paicord-rs/src/discord_gateway/user_gateway_manager.rs
use std::error::Error;
use std::sync::Arc;
use crate::discord_models::types::snowflake::Snowflake;
use crate::ffi::{self, BridgedRustError, SnowflakeRust};

use crate::discord_models::types::gateway::{GatewayEvent, PartialMessage};

#[derive(Clone)]
pub struct UserGatewayManager {
    inner: Arc<ffi::UserGatewayManager>,
}

impl UserGatewayManager {
    pub async fn new<S: AsRef<str>>(token: S) -> Self {
        let manager = ffi::user_gateway_manager_new(token.as_ref().to_string()).await;
        Self {
            inner: Arc::new(manager),
        }
    }

    pub async fn connect(&self) {
        self.inner.connect().await;
    }

    pub async fn disconnect(&self) {
        self.inner.disconnect().await;
    }

    pub async fn update_guild_subscriptions(&self, id: Snowflake, typing: bool, activities: bool, threads: bool, member_updates: bool) {
        self.inner.update_guild_subscriptions(id, typing, activities, threads, member_updates).await;
    }

    pub async fn list_messages(&self, channel_id: Snowflake, limit: i32) -> Result<Vec<PartialMessage>, Box<dyn Error + Send + Sync>> {
        match self.inner.list_messages(channel_id, limit).await {
            Ok(bridged_vec) => {
                let mut messages = Vec::new();
                for i in 0..bridged_vec.len() {
                    let bridged_message = bridged_vec.get(i);
                    messages.push(PartialMessage::new(bridged_message));
                }
                Ok(messages)
            }
            Err(e) => match e {
                BridgedRustError::UnhandledError(e) => Err(Box::new(std::io::Error::new(std::io::ErrorKind::Other, e))),
            },
        }
    }

    pub async fn request_guild_members_chunk(&self, guild_id: SnowflakeRust, user_ids: &Vec<Snowflake>) {
        self.inner.request_guild_members_chunk(guild_id, user_ids.iter().map(|id| id.get_raw()).collect()).await;
    }

    pub async fn next_event(&self) -> GatewayEvent {
        let event = self.inner.next_event().await;
        GatewayEvent::new(event)
    }
}

unsafe impl Send for ffi::UserGatewayManager {}
unsafe impl Sync for ffi::UserGatewayManager {}

File: paicord-rs/PaicordLibBridge/Sources/DiscordGateway/UserGatewayManager.swift
import Logging
import PaicordLib

func user_gateway_manager_new(token: RustString) async -> UserGatewayManager {
    return await UserGatewayManager(token: Secret(token.toString()))
}

extension UserGatewayManager {
    func next_event() async -> BridgedGatewayEvent {
        for await event: Gateway.Event in await self.events {
            return BridgedGatewayEvent(inner: event)
        }

        fatalError("Event loop stopped unexpectedly")
    }

    func update_guild_subscriptions(
        id: SnowflakeRust, typing: Bool, activities: Bool, threads: Bool, member_updates: Bool
    ) async {
        let id = GuildSnowflake(id.inner.description)

        self.updateGuildSubscriptions(
            payload:
                .init(subscriptions: [
                    id: .init(
                        typing: typing,
                        activities: activities,
                        threads: threads,
                        member_updates: member_updates,
                        channels: [:],
                        thread_member_lists: nil
                    )
                ])
        )
    }

    func list_messages(channel: SnowflakeRust, limit: Int32) async throws(BridgedRustError)
        -> BridgedMessageVec
    {
        do {
            let messages = try await client.listMessages(
                channelId: ChannelSnowflake(channel.inner),
                limit: Int(limit)
            )

            try messages.guardSuccess()

            let fetched = try messages.decode()

            return BridgedMessageVec(inner: fetched.map { BridgedPartialMessage(message: $0.toPartialMessage()) })

        } catch {
            throw BridgedRustError.UnhandledError(
                "Failed to list messages: \(error.localizedDescription)".intoRustString())
        }
    }

    func request_guild_members_chunk(guild_id: SnowflakeRust, user_ids: RustVec<UInt64>) async {
        let guildId = GuildSnowflake(guild_id.inner)
        var userIds: [UserSnowflake] = [];

        for i in 0..<user_ids.len() {
            if let userId = user_ids.get(index: UInt(i)) {
                userIds.append(UserSnowflake(userId.description))
            }
            
        }

        self.requestGuildMembersChunk(
            payload: .init(
                guild_id: guildId,
                presences: false,
                user_ids: userIds
            )
        )
    }

    func send_message(channel: SnowflakeRust, content: RustString) async {
        let nonce: MessageSnowflake = try! .makeFake(date: .now)
        do {
            try await self.client.createMessage(
                channelId: ChannelSnowflake(channel.inner),
                payload: .init(
                    content: content.toString(), nonce: .string(nonce.rawValue))
            ).guardSuccess()
        } catch {
            print("Failed to send message")
        }
    }
}

class BridgedMessageVec {
    var inner: [BridgedPartialMessage]

    init(inner: [BridgedPartialMessage]) {
        self.inner = inner
    }

    func len() -> UInt {
        return UInt(inner.count)
    }

    func get(index: UInt) -> BridgedPartialMessage {
        return inner[Int(index)]
    }
}